<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>PSoup Walkthrough • PSoup</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PSoup Walkthrough">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PSoup</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/01_introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/02_walkthrough.html">PSoup Walkthrough</a></li>
    <li><a class="dropdown-item" href="../articles/03_priorknowledge.html">Organising Prior Knowledge</a></li>
    <li><a class="dropdown-item" href="../articles/04_newtinstructions.html">Newt Editor Instructions</a></li>
    <li><a class="dropdown-item" href="../articles/05_algebraicrules.html">Algebraic Rules</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>PSoup Walkthrough</h1>
            
      

      <div class="d-none name"><code>02_walkthrough.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nicolezfortuna.github.io/PSoup/">PSoup</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: xml2</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="co">#&gt; Loading required package: prodlim</span></span>
<span><span class="co">#&gt; Loading required package: stringr</span></span>
<span><span class="co">#&gt; Loading required package: reshape2</span></span>
<span><span class="co">#&gt; Loading required package: parallel</span></span>
<span><span class="co">#&gt; Loading required package: parallelly</span></span>
<span><span class="co">#&gt; Loading required package: foreach</span></span>
<span><span class="co">#&gt; Loading required package: doParallel</span></span>
<span><span class="co">#&gt; Loading required package: iterators</span></span></code></pre></div>
<p>This is a short tutorial to demonstrate how you can use PSoup to
simulate network outcomes. Examples will use data that is included in
this package so that you can follow along even without your own
network.</p>
<div class="section level2">
<h2 id="importing-a-diagram">Importing a diagram<a class="anchor" aria-label="anchor" href="#importing-a-diagram"></a>
</h2>
<p>First start by building a diagram of your network using an editor
that can support SBGN diagrams of the Activity Flow language (I
recommend <a href="https://web.newteditor.org/" class="external-link">Newt Pathway Viewer
&amp; Editor</a>). When building your diagram, make sure that you
specify that map is of type AF, and that you only use symbols from the
Activity Flow Palette. This diagram will be based on the series of
statements that you have constructed describing your system. For advice
on how to construct your statements, view the ‘Organising Prior
Knowledge’ article. For a detailed description of how to use the Newt
editor, view the ‘Newt Editor Instructions’ article. To see the
algebraic rules that PSoup uses to convert a diagram to a mathematical
model view the ‘Algebraic Rules’ article.</p>
<center>
<div class="figure">
<img src="DunAFgenotype.jpg" alt="Figure 1. An example of a network diagram that can be produced using the Newt online editor." width="50%"><p class="caption">
Figure 1. An example of a network diagram that can be produced using the
Newt online editor.
</p>
</div>
</center>
<p>Once you are happy with your diagram, click <code>File</code> -&gt;
<code>Export</code> -&gt; <code>SBGN-ML Plain</code>. This will generate
a text file representation of your diagram using SBGN markdown language.
You will then need to use the <code>convertSBGNdiagram</code> function
to translate this text file into a Network object that is easy to use.
To use the <code>convertSBGNdiagram</code> function, you will need to
provide two arguments. The first is the file path for the .sbgn file
that describes your diagram, the second argument asks you to provide a
name for your network.</p>
<p>To trial using this function, first use the below code to import the
native PSoup network. This function will save the DunAFgenotype.sbgn
file in a location of your choosing. This function only exists to allow
the user to interact with the sbgn file for the network displayed above.
You will not need to use this function if you have your own diagram.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a folder for the example diagram to be saved in</span></span>
<span><span class="fu"><a href="../reference/getExampleDiagram.html">getExampleDiagram</a></span><span class="op">(</span>folder <span class="op">=</span> <span class="st">"~/Desktop/Diagram"</span><span class="op">)</span></span></code></pre></div>
<p>The above function will generate a Diagram folder in your Desktop
within which the DunAFgenotype.sbgn file will be saved. Using the
provided sbgn file, we can generate a network object as demonstrated
below.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a PSoup network object</span></span>
<span><span class="va">PeaNetwork</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convertSBGNdiagram.html">convertSBGNdiagram</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"~/Desktop/Diagram/DunAFgenotype.sbgn"</span>, </span>
<span>                                 networkName <span class="op">=</span> <span class="st">"PeaNetwork"</span><span class="op">)</span></span>
<span><span class="va">PeaNetwork</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; This is an object of class Network. It contains:</span></span>
<span><span class="co">#&gt; 10 hormones</span></span>
<span><span class="co">#&gt;  BranchInhibitor_S</span></span>
<span><span class="co">#&gt;  BranchOutgrowth_S</span></span>
<span><span class="co">#&gt;  BudRelease_S</span></span>
<span><span class="co">#&gt;  Cytokinin_R</span></span>
<span><span class="co">#&gt;  Feedback_R</span></span>
<span><span class="co">#&gt;  Feedback_S</span></span>
<span><span class="co">#&gt;  Inhibitor_R</span></span>
<span><span class="co">#&gt;  ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt;  Strigolactone_R</span></span>
<span><span class="co">#&gt;  Strigolactone_S</span></span>
<span><span class="co">#&gt; 5 genotypes</span></span>
<span><span class="co">#&gt;  RMS1</span></span>
<span><span class="co">#&gt;  RMS2</span></span>
<span><span class="co">#&gt;  RMS3</span></span>
<span><span class="co">#&gt;  RMS4</span></span>
<span><span class="co">#&gt;  RMS5</span></span></code></pre>
<p>Note that the hormone objects each have an underscore with wither ‘R’
or ‘S’, indicating in which compartment they are in the diagram
(‘Rootstock’ and ‘Shoot’ in this case).</p>
<p>The above network object is also natively available in the PSoup
package, and can therefore be called directly:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peaNetwork</span></span>
<span><span class="co">#&gt; This is an object of class Network. It contains:</span></span>
<span><span class="co">#&gt; 10 hormones</span></span>
<span><span class="co">#&gt;  BranchInhibitor_S</span></span>
<span><span class="co">#&gt;  BranchOutgrowth_S</span></span>
<span><span class="co">#&gt;  BudRelease_S</span></span>
<span><span class="co">#&gt;  Cytokinin_R</span></span>
<span><span class="co">#&gt;  Feedback_R</span></span>
<span><span class="co">#&gt;  Feedback_S</span></span>
<span><span class="co">#&gt;  Inhibitor_R</span></span>
<span><span class="co">#&gt;  ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt;  Strigolactone_R</span></span>
<span><span class="co">#&gt;  Strigolactone_S</span></span>
<span><span class="co">#&gt; 5 genotypes</span></span>
<span><span class="co">#&gt;  RMS1</span></span>
<span><span class="co">#&gt;  RMS2</span></span>
<span><span class="co">#&gt;  RMS3</span></span>
<span><span class="co">#&gt;  RMS4</span></span>
<span><span class="co">#&gt;  RMS5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generating-a-model-from-a-diagram">Generating a model from a diagram<a class="anchor" aria-label="anchor" href="#generating-a-model-from-a-diagram"></a>
</h2>
<p>Network objects can be passed to a function called
<code>buildModel</code> which will generate all the information that is
required to create a mathematical description of your network. You will
need to provide a directory pathway for your model information to be
stored in. This folder will be used to define your simulations, as well
as hold your simulation results. If there already exists a folder of the
specified name, this function will not overwrite the folder unless
requested to do so (you would need to pass an additional argument
<code>forceOverwrite = TRUE</code> to the buildModel function.).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define the location that all simulation material will be saved</span></span>
<span><span class="va">folder</span> <span class="op">=</span> <span class="st">"~/Desktop/Model"</span></span>
<span></span>
<span><span class="co"># generate files necessary to run simulations</span></span>
<span><span class="fu"><a href="../reference/buildModel.html">buildModel</a></span><span class="op">(</span>network <span class="op">=</span> <span class="va">PeaNetwork</span>, folder <span class="op">=</span> <span class="va">folder</span>, forceOverwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Within the folder that you have built, there will be three objects.
Two of these objects will be data frames specifying values to be fed to
the model: one will give the genotypes and their baseline values, the
second will give the starting node values for the simulation.</p>
<p>To access the information, all you need to do is load the files into
the local environment:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># To access the genotype definition object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/genotypeDef.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># To access the starting node value definition object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/nodestartDef.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genotypeDef</span></span>
<span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 1      1      1      1      1      1      1      1      1      1      1</span></span>
<span></span>
<span><span class="va">nodestartDef</span></span>
<span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 1               1           1          1           1                 1</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 1            1               1                 1          1</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 1                    1</span></span></code></pre></div>
<p>For these data frames, each row can be considered as an experimental
condition to be simulated. Data frames are initially only a single row
long, but the user can add any number of rows containing conditions that
they wish to test (more on this later). These rows can be added
manually, or they can be automatically generated using the purpose built
functions: <code>genotypeScreen</code> and
<code>randomStartScreen</code>.</p>
<p>In addition to these two data frames, the folder you generate will
also contain a script defining a function that is called
<code>nextStep</code>. This function has been automatically generated
from the PeaNetwork object based on the original diagram.</p>
<center>
<div class="figure">
<img src="buildModelOutput.png" alt="Figure 2. The nextStep function, which is automatically generated by the buildModel function." width="80%"><p class="caption">
Figure 2. The nextStep function, which is automatically generated by the
buildModel function.
</p>
</div>
</center>
<p>This function is what will be doing most of the work when running
simulations of the network, as it calculates the node values at each
step. It contains a series of difference equations (a discrete version
of a differential equation) which are the mathematical description of
the information contained in the diagram. This function has been
provided as a script accessible to the user in order to make the
translation of the provided diagram transparent to the user. For more
information on the nextStep function, view the ‘Advanced PSoup’ article
(in development).</p>
<p>Once these three data objects have been created in your specified
folder, the path directory of this folder can be provided as input to
run simulations. If we want to run a single simulation, we can call the
<code>simulateNetwork</code> function. You will notice that there is a
maxStep argument. This exists to prevent the simulation running
indefinitely in the case that the model does not find a stable state.
The simulation will also terminate in the case that one of the nodes
runs off to infinity (<code>Inf</code>).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulateNetwork.html">simulateNetwork</a></span><span class="op">(</span><span class="va">folder</span>, maxStep <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simulation</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $simulation</span></span>
<span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 2               1           1          1           1                 1</span></span>
<span><span class="co">#&gt; 3               1           1          1           1                 1</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 2            1               1                 1          1</span></span>
<span><span class="co">#&gt; 3            1               1                 1          1</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 2                    1</span></span>
<span><span class="co">#&gt; 3                    1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $stable</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre>
<p>To quickly inspect the outcome of a simulation, the
<code>fastPlot</code> function can be called as below. As the
genotypeDef and nodestartDef data frames in our model folder only
contain the baseline condition, the system is already at stability, and
therefore the plot shows no change. Note that in the plot below that the
node values have been log transformed. The <code>fastPlot</code>
function has an argument <code>logTransform</code> which by default is
set to <code>TRUE</code>. This transformation makes it easier to discern
the releative differences for positive and negative changes in the
system. If you would like to see the true node values, set this argument
to <code>FALSE</code>. This function is useful to make a quick diagnosis
of the system. To make high quality plots, use existing R plotting
tools.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fastPlot.html">fastPlot</a></span><span class="op">(</span><span class="va">simulation</span><span class="op">)</span></span></code></pre></div>
<p><img src="02_walkthrough_files/figure-html/unnamed-chunk-15-1.png" width="90%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="simulating-a-set-of-conditions">Simulating a set of conditions<a class="anchor" aria-label="anchor" href="#simulating-a-set-of-conditions"></a>
</h2>
<p>It will usually be the case that you want to test a network under
multiple conditions. Sometimes there are a few specific conditions that
you want to test, other times you want to conduct an exhaustive search.
In either case, you will need to use the <code>setupSims</code>
function, rather than the <code>simulateNetwork</code> function.</p>
<p>If there are a few specific conditions that you want to test, you can
simply add rows to the respective genotypeDef and/or nodestartDef
object. Each row will each be considered a unique condition for a
simulation. <strong>Be careful that you maintain the original column
names generated by the <code>buildModel</code> function. If they are
changed, you will be unable to run simulations.</strong></p>
<p>The code block below creates a 2nd and 3rd row, and assigns the value
1 to all available locations. When identifying a particular location in
a data.frame like your genotypeDef object, you can use square brackets
to identify the rows and columns
(<code>frameName[rows, columns]</code>). You can use this syntax even to
expand the data.frame outside of its original scope like below.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add additional rows filled with the baseline value 1</span></span>
<span><span class="va">genotypeDef</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># 2:3 is equivalent to c(2, 3). The empty column space indicates that all columns are selected</span></span>
<span></span>
<span><span class="va">genotypeDef</span></span>
<span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 1      1      1      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 2      1      1      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 3      1      1      1      1      1      1      1      1      1      1</span></span></code></pre></div>
<p>To start creating mutant conditions, the same square brackets method
can be used to indicate the value to be changed.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify mutant conditions in the additional rows</span></span>
<span><span class="va">genotypeDef</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span> <span class="co"># changing the value of the 2nd row, 2nd column to 0</span></span>
<span></span>
<span><span class="va">genotypeDef</span></span>
<span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 1      1      1      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 2      1      0      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 3      1      1      1      1      1      1      1      1      1      1</span></span></code></pre></div>
<p>Another method is to specify the column that you want to change using
<code>frameName$columnName</code>, and then just indicating the row
index</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify mutant conditions in the additional rows</span></span>
<span><span class="va">genotypeDef</span><span class="op">$</span><span class="va">RMS1_R</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span> <span class="co"># changing the value of the 'RMS1_R' column, 3rd index to 0</span></span>
<span><span class="va">genotypeDef</span><span class="op">$</span><span class="va">RMS1_S</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span> <span class="co"># changing the value of the 'RMS1_S' column, 3rd index to 0</span></span>
<span></span>
<span><span class="va">genotypeDef</span></span>
<span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 1      1      1      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 2      1      0      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 3      0      0      1      1      1      1      1      1      1      1</span></span></code></pre></div>
<p>If you want to change the values of multiple columns at once, you can
use the square bracket method again, but this time listing the names of
the columns to be changed.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify mutant conditions in the additional rows</span></span>
<span><span class="va">genotypeDef</span><span class="op">[</span><span class="fl">3</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RMS1_R'</span>, <span class="st">'RMS1_S'</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.5</span> <span class="co"># changing the value of the 3nd row, 'RMS1_R' and 'RMS1_S' columns to 0.5</span></span>
<span></span>
<span><span class="va">genotypeDef</span></span>
<span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 1    1.0    1.0      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 2    1.0    0.0      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 3    0.5    0.5      1      1      1      1      1      1      1      1</span></span></code></pre></div>
<p>In the example above, the second row represents an ‘experimental’
condition where the function of RMS2 in the scion (shoot) has been
knocked out. The third row represents a condition where the function of
RMS1 has been reduced to half of its capacity in both the root, and the
scion. Once you have generated the specific experimental conditions that
you want to test, you must save the <code>genotypeDef</code> object back
into your model folder.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save modified genotypeDef object so that it can be used for simulations</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">genotypeDef</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/genotypeDef.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<details><summary><strong>NOTE: Using Excel to define your experimental conditons</strong>
</summary><hr>
<p>In the case that you are uncomfortable interacting with a data frame
via R, you can export the data frame to excel. Once you are happy with
the conditions that you want to simulate, you must convert your xlsx
file back into an R data frame, and save it. <strong>Keep in mind that
you must not modify the column names!</strong></p>
<p>First, create an Excel readable file from the genotypeDef object in
order to preserve the original column names:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a file that can be opened in Excel</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/colearendt/xlsx" class="external-link">"xlsx"</a></span><span class="op">)</span> <span class="co"># install if you do not have this package</span></span>
<span><span class="fu">write.xlsx</span><span class="op">(</span><span class="va">genotypeDef</span>, row.names <span class="op">=</span> <span class="cn">F</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/genotypeDef.xlsx"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In Excel, add a row of 1s for every condition that you want to test.
Then modify values to correspond to the conditions that you want to
test, with each row representing one ‘experimental condition’. Leave the
first row as 1s. Once you are happy with the simulations that you want
to run, save your Excel file. You will need to load the updated Excel
file in to R, so that you can save it as an R data frame (overwriting
the original file created by buildModel).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a file that can be opened in Excel</span></span>
<span><span class="va">genotypeDef</span> <span class="op">&lt;-</span> <span class="fu">read.xlsx</span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/genotypeDef.xlsx"</span><span class="op">)</span>, sheetIndex <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">genotypeDef</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/genotypeDef.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This same process can be applied to any of the files that PSoup uses
to define simulations to be run.</p>
<hr></details><p>In the case that you want to do an exhaustive search of specifically
knockout mutations, the <code>genotypeScreen</code> function is useful
for automating the construction of the appropriate data frames. The
output of this function will automatically overwrite the genotypeDef
object in your folder, so you will not need to explicitly save your
genotypeDef object if you use this function.</p>
<p>The default behaviour of this function will create a knockout
mutation for each of the genes (modifiers) in the system, one at a time
(regardless of if they appear in multiple compartments).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/genotypeScreen.html">genotypeScreen</a></span><span class="op">(</span><span class="va">folder</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/genotypeDef.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genotypeDef</span> </span></code></pre></div>
<pre><code><span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 1      1      1      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 2      0      0      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 3      1      1      0      0      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; 4      1      1      1      1      0      0      1      1      1      1</span></span>
<span><span class="co">#&gt; 5      1      1      1      1      1      1      0      0      1      1</span></span>
<span><span class="co">#&gt; 6      1      1      1      1      1      1      1      1      0      0</span></span></code></pre>
<p>The <code>genotypeScreen</code> function can generate a data frame
with combinations of up to however many mutations that you would like to
explore, including the genotype expression values that you would like to
use for your mutations. In the case that a gene is attached to multiple
nodes in different compartments, you can choose whether to have
differential expression in the different compartments. The default
condition <code>graft = FALSE</code> indicates that all nodes modified
by the same gene should be modified identically. To allow for
differential expression in different compartments of the system, include
the argument <code>graft = TRUE</code> (eg
<code>genotypeScreen(folder, graft = TRUE) )</code>.</p>
<p>The sister function to <code>genotypeScreen</code> is
<code>randomStartScreen</code>. This function will generate a data frame
corresponding to the nodes of the network, with values indicating the
value of the node at the start of the simulation. To use this function,
you must provide the model folder path, as well as the number of restart
values that you want to generate. In addition you can also indicate what
the maximum and minimum values that you want for nodes. The default
minimum is set to 0, while the default maximum value is 2. This function
will pull values from a uniform distribution bound by these values.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/randomStartScreen.html">randomStartScreen</a></span><span class="op">(</span><span class="va">folder</span>, restarts <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/nodestartDef.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nodestartDef</span> </span></code></pre></div>
<pre><code><span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 1          1.0000      1.0000     1.0000      1.0000            1.0000</span></span>
<span><span class="co">#&gt; 2          1.5595      0.9681     0.2286      1.2914            1.6384</span></span>
<span><span class="co">#&gt; 3          0.3068      1.3820     1.6143      1.6894            1.8513</span></span>
<span><span class="co">#&gt; 4          1.9771      0.6704     1.0885      1.0960            0.4384</span></span>
<span><span class="co">#&gt; 5          0.9526      0.1929     0.9930      0.5989            0.4613</span></span>
<span><span class="co">#&gt; 6          1.6037      1.6893     1.1424      1.7921            0.6214</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 1       1.0000          1.0000            1.0000     1.0000</span></span>
<span><span class="co">#&gt; 2       1.8276          1.1447            1.2460     0.4657</span></span>
<span><span class="co">#&gt; 3       0.0760          0.9243            1.3406     1.5137</span></span>
<span><span class="co">#&gt; 4       0.8683          1.1456            0.5042     0.4706</span></span>
<span><span class="co">#&gt; 5       1.8289          0.1078            0.6614     1.2358</span></span>
<span><span class="co">#&gt; 6       0.3501          0.1684            1.0841     1.8316</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 1               1.0000</span></span>
<span><span class="co">#&gt; 2               1.5981</span></span>
<span><span class="co">#&gt; 3               1.3679</span></span>
<span><span class="co">#&gt; 4               1.4679</span></span>
<span><span class="co">#&gt; 5               1.2752</span></span>
<span><span class="co">#&gt; 6               1.1952</span></span></code></pre>
<p>The <code>modifierPriorScreen</code> function allows you to do
something similar for defining modifier values. You can learn about this
function in the ‘Advanced PSoup’ article (in development).
<code>modifierPriorScreen</code> allows for flexibility in defining the
distributions of modifier values. This functionality was designed with
the idea of being able to test the robustness of the network structure
and so has not been included in this walk through.</p>
<div class="section level3">
<h3 id="adding-an-exogenous-supply-to-a-system">Adding an exogenous supply to a system<a class="anchor" aria-label="anchor" href="#adding-an-exogenous-supply-to-a-system"></a>
</h3>
<p>In addition to being able to modify the genes and node values of the
system, it is also possible to supply an external source to a node. This
external supply is constantly given throughout the course of the
simulation, with the amount added at each step to the native value in
the simulation. If giving an exogenous supply to a series of
simulations, an <code>exogenousDef</code> object will need to be
generated and saved in the Model folder. Once this has been done, when
calling the <code>setupSims</code> function, the
<code>exogenousSupply</code> argument must be set to TRUE if you would
like to include an exogenousSupply.</p>
<p>There does not exist an automatic way to create the exogenousDef
object. You can either define the object yourself, or you can ‘cheat’ by
using the <code>nodestartDef</code> object to quickly initiate an object
that you can manipulate into an <code>exogenousDef</code> object. This
can be done because both the <code>exogenousDef</code> and
<code>nodestartDef</code> objects are defining values that relate to
nodes, and therefore their column names are defined in the same way.
<strong>NOTE that the ‘exogenousDef’ object defines how much will be
supplied to each node. 0 means nothing will be added, and is therefore
the default baseline condition.</strong></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># assign the first row of the nodestartDef to exogenousDef</span></span>
<span><span class="va">exogenousDef</span> <span class="op">&lt;-</span> <span class="va">nodestartDef</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># assign the value 0 to all locations</span></span>
<span><span class="va">exogenousDef</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="va">exogenousDef</span></span>
<span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 1               0           0          0           0                 0</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 1            0               0                 0          0</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 1                    0</span></span></code></pre></div>
<p>From this starting <code>exogenousDef</code> object, you can add rows
to define different conditions to be simulated. Specific values can be
modified in the same way as was described above for the genotypeDef
object.</p>
<details><summary><strong>NOTE: Creating a screen of exogenous supply conditions</strong>
</summary><hr>
<p>It is possible to quickly set up a full screen of two exogenous
treatments to the system by using the <code>exogenousScreen</code>
function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/exogenousScreen.html">exogenousScreen</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Strigolactone_S"</span>, <span class="st">"Cytokinin_R"</span><span class="op">)</span>, </span>
<span>                screen1 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, </span>
<span>                screen2 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, </span>
<span>                <span class="va">folder</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/exogenousDef.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">exogenousDef</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;    Strigolactone_S Cytokinin_R</span></span>
<span><span class="co">#&gt; 1                0           1</span></span>
<span><span class="co">#&gt; 2                1           1</span></span>
<span><span class="co">#&gt; 3                2           1</span></span>
<span><span class="co">#&gt; 4                3           1</span></span>
<span><span class="co">#&gt; 5                0           2</span></span>
<span><span class="co">#&gt; 6                1           2</span></span>
<span><span class="co">#&gt; 7                2           2</span></span>
<span><span class="co">#&gt; 8                3           2</span></span>
<span><span class="co">#&gt; 9                0           3</span></span>
<span><span class="co">#&gt; 10               1           3</span></span>
<span><span class="co">#&gt; 11               2           3</span></span>
<span><span class="co">#&gt; 12               3           3</span></span></code></pre>
<hr></details>
</div>
</div>
<div class="section level2">
<h2 id="executing-a-simulation-screen">Executing a simulation screen<a class="anchor" aria-label="anchor" href="#executing-a-simulation-screen"></a>
</h2>
<p>Running a simulation screen can be done automatically by using the
<code>setupSims</code> function. This function will pull data objects
from the Model folder generated by the user, to define the simulations
to be run. Users can either take a combinatorial approach to running
simulations, or they can define specifically the simulations to be
run.</p>
<p>There are a number of arguments that can be used to specify the types
of screens to be run. These arguments, which all take a logical value,
are set to <code>FALSE</code> by default.</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="80%">
</colgroup>
<thead><tr class="header">
<th>setupSims argument</th>
<th>Purpose</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>exogenousSupply</td>
<td>Specifies if the <code>exogenousDef.RData</code> object should be
imported in order to explore how the exogenous supply of resources
affect the functioning of the network.</td>
</tr>
<tr class="even">
<td>priorScreen</td>
<td>Specifies if the <code>priorDef.RData</code> should be imported in
place of the <code>genotypeDef.RData</code> in order to explore how
modifiers influence the functioning of the system when drawn from a
prior distribution. See the the ‘Advanced PSoup’ article (in
development).</td>
</tr>
<tr class="odd">
<td>robustnessTest</td>
<td>Indicates if the simulation being conducted is exploring the
behaviour of neighboring network configurations. The consequences of
this argument can be found in the ‘Advanced PSoup’ article (in
development).</td>
</tr>
<tr class="even">
<td>genotypeBaseline</td>
<td>Specifies if only the baseline condition for modifiers should be
explored regardless of the number of rows contained in the
<code>genotypeDef</code> object.</td>
</tr>
<tr class="odd">
<td>nodestartBaseline</td>
<td>Specifies if only the baseline condition for starting node values
should be explored regardless of the number of rows contained in the
<code>nodestartDef</code> object.</td>
</tr>
<tr class="even">
<td>saveOutput</td>
<td>Indicates if the simulation output should be saved in the Model
folder automatically.</td>
</tr>
<tr class="odd">
<td>combinatorial</td>
<td>In the case that multiple simulation condition objects are being
explored, this states if every combination of conditions will be
explored. If set to FALSE, every condition object will need to have the
same number of rows (unless only the baseline condition is listed). In
this instance, only corresponding rows will be simulated together.</td>
</tr>
</tbody>
</table>
<p>I know this looks very confusing, but we will go through some
examples of how this works.</p>
<p>As a consequence of the above argument defaults, the base simulation
is to explore every combination of rows contained in the
<code>genotypeDef</code> and <code>nodestartDef</code> objects. From
there, the choice can be made to also explore every combination of
exogenously supplied resource as well, or to replace the curated
<code>genotypeDef</code> object with the randomly generated
<code>priorDef</code> object. Given what the user wants to explore at
the time, the <code>genotypeDef</code> and <code>nodestartDef</code>
objects can be restricted to only the first (baseline) row being
used.</p>
<p>For the chosen combination of condition objects (eg
genotypeDef/nodestartDef objects), the default behavior of PSoup is to
simulate every combination of scenarios from each object. However, if
the user wants to run a defined set of simulations, this is possible by
setting the following argument; <code>combinatorial = FALSE</code>. In
this case, it is necessary for any screen that is not only expressing a
single condition to have the same number of rows as the same index will
be used to cycle through all screens. Therefore, each row in an explored
screen will only be simulated against the equivalent rows in other
explored screens.</p>
<div class="section level3">
<h3 id="only-exploring-the-influence-of-modifiers">Only exploring the influence of modifiers<a class="anchor" aria-label="anchor" href="#only-exploring-the-influence-of-modifiers"></a>
</h3>
<p>The following simulations will be run based on the generated
condition objects depicted above.</p>
<p>We can start by exploring the behaviour of the genotype (modifier)
values in isolation. To do this, we will need to indicate for
<code>setupSims</code> to only use the baseline values of the
<code>nodestartDef</code> object.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setupSims.html">setupSims</a></span><span class="op">(</span><span class="va">folder</span>, nodestartBaseline <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/genotypeDef_Sims.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Loading the file makes available an <code>output</code> object which
contains all simulations that were performed, including any relevant
information regarding the conditions under which the simulation was
conducted.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># quickly check the types of information contained within the output object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">output</span><span class="op">)</span></span>
<span><span class="co">#&gt;             Length Class  Mode     </span></span>
<span><span class="co">#&gt; modelFolder 1      -none- character</span></span>
<span><span class="co">#&gt; parameters  2      -none- numeric  </span></span>
<span><span class="co">#&gt; screen      6      -none- list</span></span></code></pre></div>
<p>To see the location of the Model folder where all the information
relating to the simulation was stored you can call
<code>output$modelFolder</code>. This will return the path
directory.</p>
<p>To see general parameters that have been applied to all simulations,
you can inspect the parameters. This will tell you how much delay was
applied if you used one, and what the maxStep value was.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span><span class="op">$</span><span class="va">parameters</span></span>
<span><span class="co">#&gt; maxStep   delay </span></span>
<span><span class="co">#&gt;     100       2</span></span></code></pre></div>
<p>The more interesting item to inspect is the screen. From the summary
we can see that it contains 6 lists. These are the 6 simulations that we
ran. Each simulation can be accessed by using the double brackets
(<code>output$screen[[index]]</code>).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># quickly check the types of information contained within the first simulation object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;            Length Class      Mode   </span></span>
<span><span class="co">#&gt; scenario    3     -none-     list   </span></span>
<span><span class="co">#&gt; simulation 10     data.frame list   </span></span>
<span><span class="co">#&gt; stable      1     -none-     logical</span></span></code></pre></div>
<p>Each simulation object contains three types of information. It
contains the scenario used to run the simulation, the simulation itself,
and a value which indicates if the simulation achieved stability before
the maxStep value was reached.</p>
<p>The scenario information tells you the genotype (modifier), the
starting node values, and what has been exogenously supplied (if there
was an exogenous supply). For the first screen object, every value has
the baseline condition.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">scenario</span></span>
<span><span class="co">#&gt; $genotype</span></span>
<span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 1      1      1      1      1      1      1      1      1      1      1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $startingValues</span></span>
<span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 1               1           1          1           1                 1</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 1            1               1                 1          1</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 1                    1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $exogenousSupply</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>The simulation information will show the through time calculations of
the node values, which in this case will not have changed as the first
simulation represents the baseline condition without perturbations.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">simulation</span></span>
<span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 2               1           1          1           1                 1</span></span>
<span><span class="co">#&gt; 3               1           1          1           1                 1</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 2            1               1                 1          1</span></span>
<span><span class="co">#&gt; 3            1               1                 1          1</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 2                    1</span></span>
<span><span class="co">#&gt; 3                    1</span></span></code></pre></div>
<p>Finally, the stable value is useful for seeing if stability had been
achieved without manually checking. In the baseline case, stability will
always be achieved, as the simulation starts off as stable.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">stable</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>It is possible to quickly check all of the simulations in an output
object at once to see if any of them failed to achieve stability. In
this case, all simulations reached stability.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># return a vector indicating if each simulation managed to reach stability</span></span>
<span><span class="fu"><a href="../reference/stabilityVector.html">stabilityVector</a></span><span class="op">(</span><span class="va">output</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "All simulations reached stability."</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE TRUE TRUE TRUE TRUE</span></span></code></pre></div>
<p>So far, we have only looked at the baseline conditon as the output is
much smaller. But we can use the above to inspect any of the screens
within output. For these more complicated simulations, it can be nice to
visually inspect what is going on with a plot rather than as a data
frame. Lets quickly inspect the outcome of the last simulation:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use fastPlot to quickly view the progression of a simulation</span></span>
<span><span class="fu"><a href="../reference/fastPlot.html">fastPlot</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="02_walkthrough_files/figure-html/unnamed-chunk-39-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>To see the simulation condition that was just plotted we can
call:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">scenario</span></span>
<span><span class="co">#&gt; $genotype</span></span>
<span><span class="co">#&gt;   RMS1_R RMS1_S RMS5_R RMS5_S RMS3_R RMS3_S RMS4_R RMS4_S RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 6      1      1      1      1      1      1      1      1      0      0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $startingValues</span></span>
<span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 1               1           1          1           1                 1</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 1            1               1                 1          1</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 1                    1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $exogenousSupply</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>This can be a bit clunky, and contains a lot of information that we
do not care about. In reality, all we want to know are the values that
differ from the baseline. The function <code><a href="../reference/reportCondition.html">reportCondition()</a></code>
can help us with this.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reportCondition.html">reportCondition</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; $genotype</span></span>
<span><span class="co">#&gt;   RMS2_R RMS2_S</span></span>
<span><span class="co">#&gt; 6      0      0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $startingValues</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $exogenousSupply</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>Depending on the number of conditions that have been modified, this
report might also become unweildy.</p>
</div>
<div class="section level3">
<h3 id="only-exploring-the-influence-of-the-starting-node-values">Only exploring the influence of the starting node values<a class="anchor" aria-label="anchor" href="#only-exploring-the-influence-of-the-starting-node-values"></a>
</h3>
<p>Next we can explore the influence of the starting node values in
isolation.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setupSims.html">setupSims</a></span><span class="op">(</span><span class="va">folder</span>, genotypeBaseline <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/nodestartDef_Sims.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Lets quickly inspect the outcome of one of these simulations:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reportCondition.html">reportCondition</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; $genotype</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $startingValues</span></span>
<span><span class="co">#&gt;   Strigolactone_R Inhibitor_R Feedback_R Cytokinin_R BranchOutgrowth_S</span></span>
<span><span class="co">#&gt; 2          1.5595      0.9681     0.2286      1.2914            1.6384</span></span>
<span><span class="co">#&gt;   BudRelease_S Strigolactone_S BranchInhibitor_S Feedback_S</span></span>
<span><span class="co">#&gt; 2       1.8276          1.1447             1.246     0.4657</span></span>
<span><span class="co">#&gt;   ShootDerivedSignal_S</span></span>
<span><span class="co">#&gt; 2               1.5981</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $exogenousSupply</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="fu"><a href="../reference/fastPlot.html">fastPlot</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="02_walkthrough_files/figure-html/unnamed-chunk-44-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>As you can see in the above figure, even if the starting values of
the system are different to that of the baseline, if the modifier values
are held at baseline the system will stabilise to the baseline
condition.</p>
</div>
<div class="section level3">
<h3 id="only-exploring-the-influence-of-an-exogenous-supply">Only exploring the influence of an exogenous supply<a class="anchor" aria-label="anchor" href="#only-exploring-the-influence-of-an-exogenous-supply"></a>
</h3>
<p>If we want to explore the effects of providing an exogenous supply of
resources in a baseline wildtype condition, we can do the following.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setupSims.html">setupSims</a></span><span class="op">(</span><span class="va">folder</span>, exogenousSupply <span class="op">=</span> <span class="cn">T</span>, genotypeBaseline <span class="op">=</span> <span class="cn">T</span>, nodestartBaseline <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/exogenousDef_Sims.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Lets quickly inspect the outcome of one of these simulations:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reportCondition.html">reportCondition</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; $genotype</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $startingValues</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $exogenousSupply</span></span>
<span><span class="co">#&gt;   Cytokinin_R</span></span>
<span><span class="co">#&gt; 2           1</span></span>
<span></span>
<span><span class="fu"><a href="../reference/fastPlot.html">fastPlot</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="02_walkthrough_files/figure-html/unnamed-chunk-47-1.png" width="90%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="exploring-modifiers-pulled-from-a-prior-distribution">Exploring modifiers pulled from a prior distribution<a class="anchor" aria-label="anchor" href="#exploring-modifiers-pulled-from-a-prior-distribution"></a>
</h3>
<p>The final screen condition is that of taking prior distributions of
the modifiers, here we are keeping all other possibilities as
baseline.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setupSims.html">setupSims</a></span><span class="op">(</span><span class="va">folder</span>, priorScreen <span class="op">=</span> <span class="cn">T</span>, nodestartBaseline <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">folder</span>, <span class="st">"/priorDef_Sims.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Lets quickly inspect the outcome of one of these simulations:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fastPlot.html">fastPlot</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">screen</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="02_walkthrough_files/figure-html/unnamed-chunk-50-1.png" width="90%" style="display: block; margin: auto;"></p>
<p>While only screens involving a single dimension have been
demonstrated in this tutorial, it is possible to explore multiple
dimensions at once. Use the arguments of the <code>setupSims</code>
function to specify which objects in your model folder that you want to
explore.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicole Fortuna, Christine Beveridge, ARC Centre of Excellence for Plant Success in Nature and Agriculture.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
